### PR TITLE
fix-3238 gracefully shutdown server

### LINKED ISSUE #3238: Gracefully shut down main `packages/server` process
The server package defines a [`shutdownApp`](https://github.com/medplum/medplum/blob/fbf271ba1389e850d62d154c466cf4a3cda4b173/packages/server/src/app.ts#L190) function but it appears to only get called in tests to perform clean up. We should hook into common process kill signals and clean up before shutting down the server process.

https://www.npmjs.com/package/http-graceful-shutdown might be a good option to consider this functionality.

### PR DESCRIPTION
**Pull request by dillonstreator**
Fixes #3238

### CODE REVIEW

File: packages/server/src/config.ts [hunk 1]
```diff
@@ -264,6 +265,7 @@ function addDefaults(config: MedplumServerConfig): MedplumServerConfig {
   config.botLambdaLayerName = config.botLambdaLayerName || 'medplum-bot-layer';
   config.bcryptHashSalt = config.bcryptHashSalt || 10;
   config.bullmq = { removeOnComplete: { count: 1 }, removeOnFail: { count: 1 }, ...config.bullmq };
+  config.shutdownTimeoutMilliseconds = config.shutdownTimeoutMilliseconds ?? 30000;
```
Comment (dillonstreator) [inline packages/server/src/config.ts:268]: This is redundant in its current state as the default timeout for `http-graceful-shutdown` is already 30 seconds https://github.com/sebhildebrandt/http-graceful-shutdown#options

I can remove, unless we'd like to adjust the default.

Comment (codyebberson) [reply packages/server/src/config.ts:268]: 30 seconds is appropriate.  I don't have strong feelings about whether to be explicit or implicit here.

**Comment by codyebberson** [comment]
For my own curiosity, the AWS Fargate Task timeout is also 30 seconds: https://docs.aws.amazon.com/AmazonECS/latest/APIReference/API_StopTask.html

In the common case, server shutdown should be fast (under 1 second).  The exceptions being:

1. Long running HTTP requests, such as a large FHIR `transaction` bundle
2. Long running BullMQ jobs, such as "rebuild structure definitions", "reindex resources", etc.

Those can take upwards of 30 seconds.

Overall, it feels good to be explicit.  Thanks @dillonstreator

File: packages/server/src/index.ts [hunk 2]
```diff
@@ -22,6 +23,17 @@ export async function main(configName: string): Promise<void> {
   const server = app.listen(config.port);
   server.keepAliveTimeout = config.keepAliveTimeout ?? 90000;
   globalLogger.info('Server started', { port: config.port });
+  gracefulShutdown(server, {
+    timeout: config.shutdownTimeoutMilliseconds,
+    development: process.env.NODE_ENV !== "production",
```
Comment (dillonstreator) [inline packages/server/src/index.ts:28]: @codyebberson might be nice if the config defined environment with an enum of sorts🤔

**Comment by codyebberson** [comment]
@dillonstreator  - Apologies for the inconvenience, but if you rebase & resubmit, the Github Actions problems should be fixed.  Or let me know, and we can do it on your behalf.

**Comment by dillonstreator** [comment]
> @dillonstreator - Apologies for the inconvenience, but if you rebase & resubmit, the Github Actions problems should be fixed. Or let me know, and we can do it on your behalf.

All good - I've rebased on `upstream/main` and we should now be set 👍

**Review by codyebberson — APPROVED** [review]

**Pull request merged**